src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets

```diff
@@ -0,0 +1,115 @@
+
+  <Import Project="Sdk.props" Sdk="ILLink.Tasks" />
+
+  <PropertyGroup>
+    <ILTransformDependsOn>$(ILTransformDependsOn);ILLink</ILTransformDependsOn>
+    <IntermediateLinkDir Condition=" '$(IntermediateLinkDir)' == '' ">$(IntermediateOutputPath)linked</IntermediateLinkDir>
+    <!-- Used to enable incremental build for the linker target. -->
+    <_LinkSemaphore>$(IntermediateOutputPath)Link.semaphore</_LinkSemaphore>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <LinkerRootDescriptors Include="$(LinkerRootDescriptors)" />
+  </ItemGroup>
+
+  <!--
+    ============================================================
+                     ILLink
+
+    Replace the files to be published with versions that have been
+    passed through the linker. Also prevent removed files from being
+    included in the generated deps.json.
+    ============================================================
+    -->
+  <Target Name="ILLink"
+          Condition=" '$(LinkDuringPublish)' == 'true' "
+          DependsOnTargets="RunILLink">
+
+    <!-- For now, use ResolvedFileToPublish as input/output. This
+         should go away in favor of a well-defined set of runtime
+         assemblies with https://github.com/dotnet/sdk/issues/3109. -->
+    <ItemGroup>
+      <_LinkedResolvedFileToPublish Include="@(_LinkedResolvedFileToPublishCandidates)" Condition="Exists('%(Identity)')" />
+      <ResolvedFileToPublish Remove="@(_ManagedAssembliesToLink)" />
+      <ResolvedFileToPublish Include="@(_LinkedResolvedFileToPublish)" />
+    </ItemGroup>
+
+    <!-- Remove assemblies from inputs to GenerateDepsFile. See
+         https://github.com/dotnet/sdk/pull/3086. -->
+    <ItemGroup>
+      <_RemovedManagedAssemblies Include="@(_ManagedAssembliesToLink)" Condition="!Exists('$(IntermediateLinkDir)/%(Filename)%(Extension)')" />
+
+      <ResolvedCompileFileDefinitions Remove="@(_RemovedManagedAssemblies)" />
+      <NativeCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <ResourceCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeTargetsCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimePackAsset Remove="@(_RemovedManagedAssemblies)" />
+    </ItemGroup>
+
+  </Target>
+
+
+  <!--
+    ============================================================
+                     RunILLink
+
+    Execute the linker. This target runs incrementally, only executing
+    if the output semaphore file is out of date with respect to the inputs.
+    ============================================================
+    -->
+   <UsingTask TaskName="ILLink" AssemblyFile="$(ILLinkTasksAssembly)" />
+   <Target Name="RunILLink"
+           DependsOnTargets="_ComputeManagedAssembliesToLink"
+           Inputs="@(_ManagedAssembliesToLink);@(LinkerRootDescriptors);@(ReferencePath)"
+           Outputs="$(_LinkSemaphore)">
```

Comment:
What if the set of inputs are unchanged here, but the user decides to use a different more aggressive linking mode by changing some property in the csproj. Wouldn't msbuild skip this target here, given that all inputs/outputs are unchanged?

Reply:
Good question - I see some targets that have `$(MSBuildAllProjects)` as an input. I believe that would have the desired effect. @nguerrera is that the correct approach? Is there a way to accomplish the same thing for properties passed on the command-line?

Reply:
It all depends on how careful you want to be and whether you expect scenarios to be common or not.

`$(MSBuildAllProjects)` catches changes to properties _specified in project files_. That's most projects, but it doesn't include things set on the command line. Here, that'd be something like `dotnet build -p:ExtraLinkerArgs={whatever}`.

That can also be handled, if you think it's likely to be an important scenario. Right now, it's done only for CoreCompile, added in Microsoft/msbuild#3978.

You know the scenario better, but I lean toward including `$(MSBuildAllProjects)` and leaving it at that until/unless there are significant user complaints.